File: app/src/main/java/com/stackanswer/tidakdigunakan/MovieFavoriteViewModelFactory.java
package com.stackanswer.tidakdigunakan;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class MovieFavoriteViewModelFactory  extends ViewModelProvider.NewInstanceFactory {

    private static volatile MovieFavoriteViewModelFactory INSTANCE;
    private final MovieFavoriteRepository mApplication;

    private MovieFavoriteViewModelFactory(MovieFavoriteRepository application) {
        mApplication = application;
    }

    public static MovieFavoriteViewModelFactory getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (MovieFavoriteViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new MovieFavoriteViewModelFactory(Injection.provideMovieFavriteRepository(context));
                }
            }
        }
        return INSTANCE;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(MovieFavoriteViewModel.class)) {
            return (T) new MovieFavoriteViewModel(mApplication);
        }
        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }
}

File: app/src/main/java/com/stackanswer/viewmodel/factory/kotlin/MovieFavoriteViewModelFactory.kt
package com.stackanswer.viewmodel.factory.kotlin

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.stackanswer.core.domain.usecase.MovieFavoriteUseCaseKt
import com.stackanswer.injection.kotlin.Injection
import com.stackanswer.viewmodel.MovieFavoriteViewModelKt

class MovieFavoriteViewModelFactory private constructor(private val tourismUseCase: MovieFavoriteUseCaseKt) :
    ViewModelProvider.NewInstanceFactory() {

    companion object {
        @Volatile
        private var instance: MovieFavoriteViewModelFactory? = null

        fun getInstance(context: Context): MovieFavoriteViewModelFactory =
            instance ?: synchronized(this) {
                instance ?: MovieFavoriteViewModelFactory(
                    Injection.provideMovieFavoriteUseCase(context)
                )
            }
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T =
        when {
            modelClass.isAssignableFrom(MovieFavoriteViewModelKt::class.java) -> {
                MovieFavoriteViewModelKt(tourismUseCase) as T
            }
            else -> throw Throwable("Unknown ViewModel class: " + modelClass.name)
        }
}